python/cudf/cudf/_lib/datetime.pyx

```diff
@@ -86,6 +86,33 @@ def ceil_datetime(Column col, object field):
     return result
 
 
+def floor_datetime(Column col, object field):
+    cdef unique_ptr[column] c_result
+    cdef column_view col_view = col.view()
+
+    with nogil:
+        # https://pandas.pydata.org/pandas-docs/version/0.25.0/reference/api/pandas.Timedelta.resolution.html
+        if field == "D":
+            c_result = move(libcudf_datetime.floor_day(col_view))
+        elif field == "H":
+            c_result = move(libcudf_datetime.floor_hour(col_view))
+        elif field == "T":
+            c_result = move(libcudf_datetime.floor_minute(col_view))
+        elif field == "S":
+            c_result = move(libcudf_datetime.floor_second(col_view))
+        elif field == "L":
+            c_result = move(libcudf_datetime.floor_millisecond(col_view))
+        elif field == "U":
+            c_result = move(libcudf_datetime.floor_microsecond(col_view))
+        elif field == "N":
+            c_result = move(libcudf_datetime.floor_nanosecond(col_view))
+        else:
+            raise ValueError(f"Invalid resolution: '{field}'")
+
+    result = Column.from_unique_ptr(move(c_result))
```

Comment:
It looks like the `move` is done twice, once when `c_result` is set and again here. Is calling `move` twice necessary for the intended move semantics? I think it might be fine to remove the first `move` and leave it in this line. I'm not 100% sure about this.

Reply:
Yes -- the C++ code that Cython generates unfortunately necessitates these calls to `move()` when working with movable C++ types. I believe you'll get a (C++) compile error if  you  remove  either one of them.

Reply:
Thanks @shwina! I haven't dealt with Cython move semantics yet so I appreciate the info.

Reply:
The first move is required because the C++ function call exposed via Cython is decorated with a `noexcept +`. The Cython-generated exception handling logic involves assigning the result of the C++ API to a temporary followed by an inspection of the exception payload, and the result is only assigned from the temporary if no exception was encountered. If you skip the first move the `result = tmp` assignment becomes an attempt to copy rather than move the `unique_ptr` and therefore fails.

Reply:
> If you skip the first move the result = tmp assignment becomes an attempt to copy rather than move the unique_ptr and therefore fails.

@vyasr can you explain more with another example? if a `noexcept +` is not used, then what @bdice said could work essentially?

Reply:
Yes, basically I _think_ that should work, although I'd have to test it to be sure that the Cython-generated C++ still avoids the temporary assignment when not catching an exception. I'm not sure how much you know about move semantics in C++ to give a suitable example, but basically the problem is this:
```
auto x = std::make_unique<...>(...);
auto y = x;  <-- This is illegal because it violates the "unique" part of a unique_ptr.
auto y = std::move(x);  <-- This basically says "x no longer owns anything". It's legal, but you can't use x after.
```

When you expose a C++ function via Cython and use `except +`, Cython generates C++ that looks (very roughly) like this:
```
try {
    tmp = do_something(...);
} catch (...) {
    // Handle exception if necessary
}
result = tmp;
```
If `do_something` returns a unique_ptr, the second assignment `result=tmp` is invalid. You need that to be `result=move(tmp)`. That's what adding the first move call does.

Reply:
I'll just add to Vyas' excellent explanation of the problem that we definitely want to keep the `except +`. Annotating Cython functions with `except +` ensure that any C++ exceptions raised within those functions are translated into an (appropriate) Python `Exception`.

As an educational exercise, I'd highly recommend inspecting the C++ code generated by Cython both with and without `except +`, and that should help clarify why the `move()` is not necessary without `except +`.